src/test/software/amazon/event/ruler/MachineTest.java

```diff
@@ -2347,7 +2347,7 @@ public void testApproximateSizeForRulesManySouceAndEventNameArrayElements() thro
                         "        \"eventName\": [\"Name6\",\"Name7\",\"Name8\",\"Name9\",\"Name10\"]\n" +
                         "    }\n" +
                         "}");
-        assertEquals(90, machine.approximateObjectCount());
+        assertEquals(79, machine.approximateObjectCount());
     }
 
 }
```

Comment:
Suggest to add a test to ensure the name state reuse among different rules and it doesn't matter the sequence any more, the same name states should be always reused.

For example:
```
        machine.addRule("rule1", "{\"bar\":[\"x\", \"y\"]}");
        machine.addRule("rule2", "{\"bar\":[\"x\"]}");
        machine.addRule("rule3", "{\"bar\":[\"y\"]}");
```
yields one new name state, and below will totally reuse the name state created above without any new name state being created ...
```
        machine.addRule("rule2", "{\"bar\":[\"x\"]}");
        machine.addRule("rule3", "{\"bar\":[\"y\"]}");
        machine.addRule("rule1", "{\"bar\":[\"x\", \"y\"]}");
```
assert only one name state being created overall.
also verify the deletion does also care about the delete sequence, the new namestate will only be removed when the last key have been removed from state.

Reply:
Done